### ISSUE TITLE
Lemma with an `ensures` clause is not being applied

### ISSUE THREAD

**Issue by lucasmt** [runtimeverification/haskell-backend#3669]
We are using the following lemma:
```
rule [keccak-eq-conc-false]: keccak(A)  ==Int B => false
    ensures #notEq(keccak(A), B, "keccak-eq-conc-false") [symbolic(A), concrete(B), simplification, comm]
```
The idea is that sometimes we want to assume that the `keccak` of a symbolic expression will not collide with a pre-existing concrete value, but we are using the `ensures` clause to record this assumption in the configuration (with a new `#notEq` symbol) so we can check afterwards that the cases where this lemma is applied are sound.

However, we have the following claim:

```
claim [lookup-keccak]:
    <k> runLemma ( #lookup ( STORAGE:Map [ 1 <- 2 ] , keccak ( X ) ) )
    => doneLemma ( #lookup ( STORAGE:Map , keccak ( X ) ) ) ... </k>
```
It should be passing by the lemma above and [this lemma](https://github.com/runtimeverification/evm-semantics/blob/c0e85adece461201de15e637a2bb002a026929c4/kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/lemmas.k#L84) about `#lookup`, but it fails with the lemma as it is. However, if we remove the `ensures` clause, it passes.

I'm attaching a `keccak-lemmas.k` file with the lemma and claim above. The behavior should be reproducible using `kprove`. It's worth noting that this claim is based on a proof that we were running with the booster, where this subexpression was also not being simplified.

[keccak-lemmas.k.txt](https://github.com/runtimeverification/haskell-backend/files/12796833/keccak-lemmas.k.txt)

**Comment by lucasmt** [comment]
I mentioned this issue to @goodlyrottenapple in our meeting earlier today.

**Comment by ehildenb** [comment]
Would it be sufficient to log it whenever this equation applies, then inspect the log after the proof is finished?

Rather than building it into the ensure clause, basically make it meta-level feature (which it is), by just telling the backend to log applications of this rule?

**Comment by lucasmt** [comment]
@ehildenb Can we have the backend log applications of certain rules? Yeah, if we can do that we wouldn't need the `ensures` clause.

**Comment by ehildenb** [comment]
I think that should be an easy enough feature to implement, yes. I think it already can do that actually, not sure it will also log the substitution it applies, but we can tell it to log rules applied, and then pyk can extract those logs.

I guess it wolud be nice to tell it to only log the specific rule we care about, because otherwise th elogs will be huge and the backend much slowed down.

@geo2a is it possible? Does this information exist in the logs? Can we tell (over RPC), to only log specific equations by name, and also to give us the sustitution that equation applied with in the log?

**Comment by goodlyrottenapple** [comment]
At the moment the only option is `"log-successful-simplifications": true` which will return entries of this shape
```json
{ 
  "tag": "simplification",
  "original-term": ...,
  "result": {
    "tag": "success",
    "rewritten-term": ...,
    "rule-id": ...
  }
}
```
perhaps the easiest thing would be to modify the `"log-successful-simplifications": true` parameter to either take a bool or a list of rule labels, i.e. `"log-successful-simplifications": ["keccak-eq-conc-false"]`?

**Comment by ehildenb** [comment]
Yeah, that would be great. And I guess the `original-term` field contains the needed information about what the arguments to `keccak` were.

**Comment by geo2a** [comment]
Out experience shows that simplification logs very quickly become prohibitively large (10s of GB over one request), therefore we will implement a more specialized version, as @goodlyrottenapple suggested. @ehildenb @lucasmt do I understand correctly that we need to log the arguments too, not only the tag? Could you come up with the minimal amount of information we need to log? How many applications of the "tracked rule" do you envision during one proof?

**Comment by lucasmt** [comment]
@geo2a Yeah, we would want to know the arguments as well. That way if we ever have an application of the rule that looks like this, for example:
```
keccak(A)  ==Int keccak(b"") => false
```
we would be able to catch it. So ideally we'd like to log the subexpression that matches the left-hand side of the rule, or else the substitutions (like `A |-> A, B |-> keccak(b"")`). I don't actually know how many applications we would expect, because they happen behind the scenes. Maybe @PetarMax has some idea?
